ewms_pilot/pilot.py

```diff
@@ -144,41 +167,105 @@ def read_from_subproc(
     return out_msg
 
 
-def process_msg(
+async def _stream_to_file(stream: Optional[asyncio.StreamReader], fpath: Path) -> None:
+    if not stream:
+        return
+    with open(fpath, "w+") as f:
+        while not stream.at_eof():
+            data = await stream.readline()
+            if not data:
+                continue
+            line = data.decode("ascii").rstrip()
+            print(line, file=f)
+
+
+async def process_msg_task(
     in_msg: Any,
     cmd: str,
     subproc_timeout: Optional[int],
-    fpath_to_subproc: Path,
-    fpath_from_subproc: Path,
+    #
+    ftype_to_subproc: FileType,
+    ftype_from_subproc: FileType,
+    #
     file_writer: Callable[[Any, Path], None],
     file_reader: Callable[[Path], Any],
+    #
     debug_dir: Optional[Path],
+    pub: mq.queue.QueuePubResource,
 ) -> Any:
-    """Process the message in a subprocess using `cmd`."""
+    """Process the message's task in a subprocess using `cmd` & respond."""
+    task_id = str(time.time())
+
     # debugging logic
     debug_subdir = None
+    debug_file_stdout = None
+    debug_file_stderr = None
     if debug_dir:
-        debug_subdir = debug_dir / str(time.time())
+        debug_subdir = debug_dir / task_id
         debug_subdir.mkdir(parents=True, exist_ok=False)
+        debug_file_stdout = debug_subdir / "stdout"
+        debug_file_stderr = debug_subdir / "stderr"
+
+    # create in/out filepaths
+    fpath_to_subproc = Path(f"in-{task_id}{ftype_to_subproc.value}")
+    fpath_from_subproc = Path(f"out-{task_id}{ftype_from_subproc.value}")
+
+    # insert in/out files into cmd
+    cmd = cmd.replace("{{INFILE}}", str(fpath_to_subproc))
+    cmd = cmd.replace("{{OUTFILE}}", str(fpath_from_subproc))
 
     # write
     write_to_subproc(fpath_to_subproc, in_msg, debug_subdir, file_writer)
 
     # call & check outputs
     LOGGER.info(f"Executing: {shlex.split(cmd)}")
     try:
-        subprocess.run(
-            shlex.split(cmd),
-            check=True,
+        # await to start
+        proc = await asyncio.create_subprocess_shell(
+            cmd,
+            stdout=asyncio.subprocess.PIPE,
+            stderr=asyncio.subprocess.PIPE,
+        )
+
+        # await to finish while streaming output
+        # > prep
+        coros: List[Coroutine[Any, Any, Any]] = [proc.wait()]
+        if debug_file_stdout:
+            coros.append(_stream_to_file(proc.stdout, debug_file_stdout))
+        if debug_file_stderr:
+            coros.append(_stream_to_file(proc.stdout, debug_file_stderr))
```

Comment:
If you're using `PIPE`, these can't be optional.  Otherwise the `wait()` will deadlock with larger outputs.

I'd also recommend that if you're going to stream to file, you provide the file handle directly to the subprocess.  That will definitely eliminate any deadlocks.  For example:

```python
if debug:
    stdoutf = open('outfile', 'wb')
    stderrf = open('errfile', 'wb')
    proc = await asyncio.create_subprocess_shell(
        cmd,
        stdout=stdoutf,
        stderr=stderrf
    )
else:
    proc = await asyncio.create_subprocess_shell(cmd)
```

Reply:
thanks, see [d95ec21](https://github.com/Observation-Management-Service/ewms-pilot/pull/28/commits/d95ec21d36f3585c782dc7bf54ef6c8fb19d2ba8)

Reply:
Better, but I still wonder why `_stream_to_file()` is even being used.  With the code as-is, it just seems like an unnecessary complication instead of letting the subprocess write directly to the file(s).

Reply:
See my example code above, which passes the file handles directly to the subprocess, instead of using a pipe.

Reply:
Aha! I misread your example yesterday